src/compiler/scala/tools/nsc/Reporting.scala

```diff
@@ -142,3 +270,334 @@ trait Reporting extends internal.Reporting { self: ast.Positions with Compilatio
     }
   }
 }
+
+object Reporting {
+  sealed trait Message {
+    def pos: Position
+    def msg: String
+    def category: WarningCategory
+    def site: String // sym.FullName of the location where the warning is positioned, may be empty
+  }
+
+  object Message {
+    final case class Plain(pos: Position, msg: String, category: WarningCategory, site: String) extends Message
+
+    // `site` and `origin` may be empty
+    final case class Deprecation(pos: Position, msg: String, site: String, origin: String, since: Version) extends Message {
+      def category: WarningCategory = WarningCategory.Deprecation
+    }
+  }
+
+  sealed trait WarningCategory {
+    lazy val name: String = {
+      val objectName = this.getClass.getName.split('$').last
+      WarningCategory.insertDash.replaceAllIn(objectName, "-")
+        .stripPrefix("-")
+        .stripSuffix("-")
+        .toLowerCase
+    }
+
+    def includes(o: WarningCategory): Boolean = this eq o
+    def summaryCategory: WarningCategory = this
+  }
+
+  object WarningCategory {
+    private val insertDash = "(?=[A-Z][a-z])".r
+
+    var all: mutable.Map[String, WarningCategory] = mutable.Map.empty
+    private def add(c: WarningCategory): Unit = all += ((c.name, c))
+
+    object Deprecation extends WarningCategory; add(Deprecation)
+
+    object Unchecked extends WarningCategory; add(Unchecked)
+
+    object Optimizer extends WarningCategory; add(Optimizer)
+
+    object Scaladoc extends WarningCategory; add(Scaladoc)
+
+    object JavaSource extends WarningCategory; add(JavaSource)
+
+    sealed trait Other extends WarningCategory { override def summaryCategory: WarningCategory = Other }
+    object Other extends Other { override def includes(o: WarningCategory): Boolean = o.isInstanceOf[Other] }; add(Other)
+    object OtherShadowing extends Other; add(OtherShadowing)
+    object OtherPureStatement extends Other; add(OtherPureStatement)
+    object OtherMigration extends Other; add(OtherMigration)
+    object OtherMatchAnalysis extends WarningCategory; add(OtherMatchAnalysis)
+    object OtherDebug extends WarningCategory; add(OtherDebug)
+
+    sealed trait WFlag extends WarningCategory { override def summaryCategory: WarningCategory = WFlag }
+    object WFlag extends WFlag { override def includes(o: WarningCategory): Boolean = o.isInstanceOf[WFlag] }; add(WFlag)
+    object WFlagDeadCode extends WFlag; add(WFlagDeadCode)
+    object WFlagExtraImplicit extends WFlag; add(WFlagExtraImplicit)
+    object WFlagNumericWiden extends WFlag; add(WFlagNumericWiden)
+    object WFlagOctalLiteral extends WFlag; add(WFlagOctalLiteral)
```

Comment:
This is the one that caught my attention, as it is a very narrow warning. And well-meaning if misguided contributors add annoying warnings all the time that I probably want to suppress because they are so annoying.

Reply:
This is the best I could come up with.. We cannot put `add(this)` it in the constructor of `WarningCategory`, as the objects here are evaluated lazily.